server/streamreader/jack_stream.cpp

```diff
@@ -0,0 +1,460 @@
+            interleave_func_ = float_to_s24;
+            break;
+        case 32:
+            interleave_func_ = float_to_s32;
+            break;
+        default:
+            throw SnapException("JackStreams only support 16, 24 and 32 bit sample formats");
+    }
+
+#if 0
+    // Disabled by default as aixlog <= 1.4.0 is not thread-safe at the moment
+    // and Jack produces quite a few log message.
+    // TODO: enable if and when aixlog becomes thread-safe.
+    jack_set_error_function([](const char* msg) { LOG(ERROR, LOG_TAG) << "Jack Error: " << msg << "\n"; });
+    jack_set_info_function([](const char* msg) { LOG(DEBUG, LOG_TAG) << msg << "\n"; });
+#endif
+}
+
+
+void JackStream::start()
+{
+    LOG(TRACE, LOG_TAG) << "JackStream::start()\n";
+
+    // Need to start immediately, otherwise client will fail
+    // due to a zero-length pcm header message
+    tvEncodedChunk_ = std::chrono::steady_clock::now();
+    PcmStream::start();
+
+    boost::asio::post(strand_, [this] { tryConnect(); });
+}
+
+
+void JackStream::stop()
+{
+    LOG(TRACE, LOG_TAG) << "JackStream::stop()\n";
+
+    PcmStream::stop();
+    closeJackConnection();
+}
+
+
+/**
+ * Try to connect to the jack server. If it fails, wait
+ * a little bit and try again.
+ */
+void JackStream::tryConnect()
+{
+    try
+    {
+        if (!openJackConnection())
+        {
+            LOG(WARNING, LOG_TAG) << "Jack connection failed, trying again in 5 seconds\n";
+            wait(read_timer_, 5s, [this] { tryConnect(); });
+            return;
+        }
+
+        if (!createJackPorts())
+        {
+            LOG(ERROR, LOG_TAG) << "Failed to create Jack ports, trying again in 5 seconds\n";
+            closeJackConnection();
+            wait(read_timer_, 5s, [this] { tryConnect(); });
+            return;
+        }
+
+        PcmStream::start();
+    }
+    catch (exception& e)
+    {
+        LOG(ERROR, LOG_TAG) << "Error during Jack connection: " << e.what() << "\n";
+        stop();
+    }
+}
+
+
+bool JackStream::openJackConnection()
+{
+    char* serverName = serverName_.data();
+    jack_options_t options = (jack_options_t)(JackNoStartServer | JackServerName);
+
+    client_ = jack_client_open(name_.c_str(), options, nullptr, serverName);
+    if (client_ == nullptr)
+    {
+        return false;
+    }
+
+    LOG(INFO, LOG_TAG) << "Connected Jack client " << jack_get_client_name(client_) << "\n";
+
+    jack_nframes_t jack_sample_rate = jack_get_sample_rate(client_);
+    if (jack_sample_rate != sampleFormat_.rate())
+    {
+        throw SnapException("Jack streams must match the sample rate of the Jack server. "
+                            "The server sample rate is " +
+                            cpt::to_string(jack_sample_rate) + ".");
+    }
+
+    jack_time_t jackTime = jack_get_time();
+    auto now = std::chrono::steady_clock::now();
+    jackTimeAdjust_ = chrono::duration_cast<chrono::microseconds>(now.time_since_epoch()).count() - jackTime;
```

Comment:
@badaix I'm not entirely happy with this time-adjustment. I think the `std::chrono::steady_clock` and Jacks timing source is not going to move apart during runtime, but it still feels a little hackish (and is not 100% accurate, of course). I thought about modifying SnapCast so that multiple (named) clock sources could be used. Then each stream could optionally specify which clock source to use for synchronization. It seemed to invasive to me at first, but having now read more of SnapCast sources makes me think that it might be a fairly easy (and probably even backwards compatible) change. But I would do that in a separate PR, if you think it might be worth pursuing.

Reply:
I have to take a deeper look at this and test it. Current review comments were more like a "syntactical linting" to clean up and silence compiler warnings and static analysis. I also stumbled upon this, also in `configuration.md`. I didn't understand what this means:
> The timing information is taken from the Jack server, so all
streams connected to the same Jack server should play perfectly in sync.

Is it about jack stream `playback_1` being in sync with `playback_2` or is it about Snapclient A plays `playback_1` in sync with Snaplclient B, playing also `playback_1`?
The latter case Snapcast is taking care of. 

This adjustment looks dangerous, because it will in the end alter the total delay (the one which defaults to 1000ms), and might lower it to 5ms, causing clients to not being able to play anything, because the received signal seems to be too old to be played out.

Reply:
> Is it about jack stream `playback_1` being in sync with `playback_2` or is it about Snapclient A plays `playback_1` in sync with Snaplclient B, playing also `playback_1`?
> The latter case Snapcast is taking care of.

I know SnapCast is taking care of the latter case, and it does so very well! That's actually the reason I decided to use SnapCast for my project. So it's a variation of the first case: Client A, Stream 1 should be in sync with client B, stream 2 with client C, stream 3.

> This adjustment looks dangerous, because it will in the end alter the total delay (the one which defaults to 1000ms), and might lower it to 5ms, causing clients to not being able to play anything, because the received signal seems to be too old to be played out.

I don't think it will do that, as all the other streamreaders just set the chunk timestamp to `::now()` as well. The Jack time adjustment does just that: adjust the Jack timestamp to be in line with the other streamreaders. But the good part for my project is: all Jack streamreaders use exactly the same timestamp for chunks which come from the same Jack timeslice. So if I have a 24 channel Jack output and split them into 24 individual 48000:16:1 streams, all streams are perfectly in sync.

Reply:
But I might have misunderstood the tvEncodedChunk / timing stuff. I thougt the playback delay is added later, when sending the chunk to the clients?

Reply:
The `tvEncodedChunk` must be initialized once on every playback start. The `chunkRead` callback will add the chunk's duration to the timestamp, so if 20ms are read, it will be incremented by 20ms. 

You can "see" how good the timestamps are by enabling debug log on the client:

```
snapclient --logfilter=*:debug -h 127.0.0.1
...
2024-06-06 20-27-11.999 [Debug] (Player) setVolume exp with base 10: 0.45 => 0.202043
2024-06-06 20-27-11.999 [Debug] (Alsa) Resizing buffer from 0 to 28220
2024-06-06 20-27-11.999 [Info] (Stream) No chunks available
2024-06-06 20-27-11.999 [Info] (Alsa) Failed to get chunk
2024-06-06 20-27-12.020 [Info] (Controller) diff to server [ms]: -0.009
2024-06-06 20-27-12.373 [Debug] (Stream) Silent frames: 41, frames: 902, age: -0.94
2024-06-06 20-27-12.393 [Debug] (Stats) Chunk: 0	0	0	0	1	80	0
2024-06-06 20-27-13.007 [Debug] (Stats) Chunk: 0	0	0	0	31	80	0
2024-06-06 20-27-14.015 [Debug] (Stats) Chunk: -1	-1	0	0	81	80	0
2024-06-06 20-27-15.009 [Debug] (Stats) Chunk: -3	-3	-1	-1	130	80	0
2024-06-06 20-27-16.013 [Debug] (Stats) Chunk: -4	-3	-3	-2	180	80	3
2024-06-06 20-27-17.022 [Debug] (Stats) Chunk: -2	-2	-3	-2	230	79	8
2024-06-06 20-27-18.023 [Debug] (Stats) Chunk: -1	-1	-2	-2	279	79	8
2024-06-06 20-27-19.010 [Debug] (Stats) Chunk: -1	-1	-1	-2	328	80	6
2024-06-06 20-27-20.004 [Debug] (Stats) Chunk: 0	0	-1	-1	377	80	3
2024-06-06 20-27-21.013 [Debug] (Stats) Chunk: 0	0	0	-1	427	80	3
2024-06-06 20-27-22.021 [Debug] (Stats) Chunk: -1	-1	-1	-1	477	79	0
2024-06-06 20-27-23.015 [Debug] (Stats) Chunk: -1	-1	-1	-1	500	80	2
2024-06-06 20-27-24.014 [Debug] (Stats) Chunk: -1	-1	-1	-1	500	79	4
2024-06-06 20-27-25.021 [Debug] (Stats) Chunk: -1	-1	-1	-1	500	80	3
2024-06-06 20-27-26.019 [Debug] (Stats) Chunk: -1	-1	-1	-1	500	80	4
^C2024-06-06 20-27-26.860 [Info] (Snapclient) Received signal 2: Interrupt
2024-06-06 20-27-26.867 [Debug] (Connection) Disconnecting
2024-06-06 20-27-26.867 [Debug] (Connection) Disconnected
2024-06-06 20-27-26.867 [Notice] (Snapclient) Snapclient terminated.
```

The closer the values are to 0 in the `(Stats) Chunk` logs, the better. These logs are taken from a modified `jack_stream.cpp` where I've removed manual correction of the timestamp.
The client log above looks fine, the values might change from DAC to DAC and depend on the load of the system and the temperature of the DAC, ... The last number are the inserted/removed samples per second to keep up sync.

Reply:
My approach to use Jack as a timing source for the chunk timestamping was quite flawed. Main problem, apart from me not understanding how `tvEncodedChunk` is used at first, is that Jack uses CLOCK_MONOTONIC_RAW on Linux by default, SnapCast uses CLOCK_MONOTONIC (through `std::chrono:steady_clock`).

I will think more about how to achieve my goal of having identical timestamps for all streams of a single Jack server. Maybe I will do it as a small patch (to Jack and SnapCast) or I might send another PR if I think it's clean enough to be used by others.